### ISSUE TITLE
Refactor APIs for unbound streaming (attach, log and events commands)

### ISSUE THREAD

**Issue by marcuslinke** [docker-java/docker-java#231]
The problem is with unbound streaming thus using log command 'followStream' option or using events command without 'until' option. If the server returns an unbound stream normally there is no chance to cancel the stream as `InputStream#read()' blocks while waiting for chunks to come from the server. This can be solved only by using separate threads as in the current events command implementation.

I'm still unsatisfied with these different streaming APIs in docker-java. I think all streaming commands should use the same callback driven API as in events command.

**Comment by bryanstephens** [comment]
I'm currently seeing an issue with the events streaming api in that if I try to shutdown the ExecutorService that is handling the stream then BufferedInputStream#close() is called and it blocks indefinitely trying to finish reading from the socket. 

Investigating this I found that when close is called in the finally block of the EventNotifier that the FIN flag is not being sent to the docker tcp/unix socket therefor leaving the BufferedInputStream waiting for an EOF that will never come.

I'm not sure if this is tied to what is described above or if I should open a separate ticket for this.

**Comment by marcuslinke** [comment]
There should be no such problem anymore as  https://github.com/docker-java/docker-java/commit/a072d62104d03d53cf2f9a41b95fd527fb6acbb9 has fixed this. Please ensure that you have latest docker-java version in place. Feel free to open a new issue if your problem still persist.

**Comment by bryanstephens** [comment]
@marcuslinke Is there a plan to have a released build that includes that commit? I will build and test locally but I would rather pull down an official dependency. I'm pulling based on the latest tag I see here which is 1.3.0 and I was having this issue.

**Comment by marcuslinke** [comment]
Please test with latest SNAPSHOT. If this solves your issue i will build a new release.

**Comment by marcust** [comment]
At least for me a current SNAPSHOT solves all issues with unclosed streams.

**Comment by bryanstephens** [comment]
It isn't the cleanest shutdown, it actually still hangs there until another event comes along and then it shuts down. You can see this happen if you await termination on the executor service, and make sure no new events are triggered, it will timeout. However if you are awaiting termination and trigger another event it no longer times out. I can make due with it though.

**Comment by marcuslinke** [comment]
@bryanstephens Could you post some example code please?

**Comment by bryanstephens** [comment]
@marcuslinke with this gist: https://gist.github.com/bryanstephens/5d4e802c2be1e49ff661

I'm running this off of a snapshot build at commit ab2eabf29e8fa3c56bfc5c55e1cfe8ea0f035fe8.

I still see the thread never does shutdown. This is the output I see:

```
The executor service is still running, I'm just continuing to run because I timed out.
executorServiceThatWillStillBeBlocking.isShutdown() = true, executorServiceThatWillStillBeBlocking.isTerminated() = false
```

As you can see the executor service has been triggered to shutdown but it has yet to get terminated. This is the current state of the thread:

```
"pool-1-thread-1" #12 prio=5 os_prio=31 tid=0x00007fcd730d0800 nid=0x5603 runnable [0x0000000129a74000]
   java.lang.Thread.State: RUNNABLE
	at java.net.SocketInputStream.socketRead0(Native Method)
	at java.net.SocketInputStream.socketRead(SocketInputStream.java:116)
	at java.net.SocketInputStream.read(SocketInputStream.java:170)
	at java.net.SocketInputStream.read(SocketInputStream.java:141)
	at org.apache.http.impl.io.SessionInputBufferImpl.streamRead(SessionInputBufferImpl.java:136)
	at org.apache.http.impl.io.SessionInputBufferImpl.fillBuffer(SessionInputBufferImpl.java:152)
	at org.apache.http.impl.io.SessionInputBufferImpl.readLine(SessionInputBufferImpl.java:270)
	at org.apache.http.impl.io.ChunkedInputStream.getChunkSize(ChunkedInputStream.java:240)
	at org.apache.http.impl.io.ChunkedInputStream.nextChunk(ChunkedInputStream.java:206)
	at org.apache.http.impl.io.ChunkedInputStream.read(ChunkedInputStream.java:169)
	at org.apache.http.conn.EofSensorInputStream.read(EofSensorInputStream.java:137)
	at java.io.BufferedInputStream.fill(BufferedInputStream.java:246)
	at java.io.BufferedInputStream.read(BufferedInputStream.java:265)
	- locked <0x000000076c050938> (a java.io.BufferedInputStream)
	at java.io.FilterInputStream.read(FilterInputStream.java:83)
	at org.glassfish.jersey.message.internal.EntityInputStream.isEmpty(EntityInputStream.java:185)
	at org.glassfish.jersey.message.internal.InboundMessageContext.hasEntity(InboundMessageContext.java:747)
	at com.github.dockerjava.core.util.LoggingFilter.filter(LoggingFilter.java:251)
	at org.glassfish.jersey.client.ClientFilteringStages$ResponseFilterStage.apply(ClientFilteringStages.java:134)
	at org.glassfish.jersey.client.ClientFilteringStages$ResponseFilterStage.apply(ClientFilteringStages.java:123)
	at org.glassfish.jersey.process.internal.Stages.process(Stages.java:171)
	at org.glassfish.jersey.client.ClientRuntime.invoke(ClientRuntime.java:251)
	at org.glassfish.jersey.client.JerseyInvocation$2.call(JerseyInvocation.java:683)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:315)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:297)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:228)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:424)
	at org.glassfish.jersey.client.JerseyInvocation.invoke(JerseyInvocation.java:679)
	at org.glassfish.jersey.client.JerseyInvocation$Builder.method(JerseyInvocation.java:408)
	at org.glassfish.jersey.client.JerseyInvocation$Builder.get(JerseyInvocation.java:308)
	at com.github.dockerjava.jaxrs.EventsCmdExec$EventNotifier.call(EventsCmdExec.java:74)
	at com.github.dockerjava.jaxrs.EventsCmdExec$EventNotifier.call(EventsCmdExec.java:50)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

**Comment by marcuslinke** [comment]
I think the cause of the problem is that the callback's `isReceiving` method still returns true. Please adopt your code according to https://github.com/docker-java/docker-java/blob/master/src/test/java/com/github/dockerjava/core/command/EventsCmdImplTest.java#L75 

The events API is something that needs some refactoring to be more clear. Thats something I'm currently working on. Thanks for investigating anyway.

**Comment by bryanstephens** [comment]
Ah yes, in my actual project I'm actually setting it to false when shutting down which is where I am seeing the weird behavior in that it would only shutdown once another event came through.

**Comment by marcuslinke** [comment]
OK, now i got it. With the current API the callback has no chance to close the underlying `InputStream` / `Response`. So it only signals that is is not interested anymore in consuming the stream. This can be evaluated by the callback-driving thread with the next event steamed from the server only as the read blocks until. Hope I'm able to merge the API changes within the next few days.

**Comment by bryanstephens** [comment]
Just curious what the progress looks like for this. If you need some help I can help out. I'm not much of an expert with Jersey and the Apache Http Client, I've learned enough to be dangerous with my fork to use Jersey 1.11 as docker-java had a conflicting dependency with other dependencies on my project. It appears that I can now use docker-java as is with the exception of this one issue which is also present in my fork.

**Comment by marcuslinke** [comment]
I've pushed a new branch `refact-streaming` (https://github.com/docker-java/docker-java/tree/refact-streaming) that contains the appropriate API changes. I would like to collect some input / opinions before merge it into `master` branch. So feel free to review.

There is one problem that I can't resolve. The attach command doesn't behave as described in the docker remote api reference. The response stream coming from the docker host seems to be NOT multiplexed. Don't know if this is a problem of the underlying `HttpClient`...

@alexec What do you think about these API changes?

**Comment by gesellix** [comment]
@marcuslinke whether the stream is multiplexed or not depends on the container config. You need to ensure that `container.Config.Tty == false`. That was at least my experience when working with the remote api :-)

**Comment by docker-java-maintain** [comment]
Yes i know but it doesnt behave like this .
Am 27.06.2015 11:25 schrieb "Tobias Gesellchen" <notifications@github.com>:

**Comment by bryanstephens** [comment]
@marcuslinke So far from trying to use the refactored api I can tell you it seems like ResultCallbackTemplate should be an abstract class.

**Comment by bryanstephens** [comment]
@marcuslinke I am still getting into a situation where the events thread isn't properly shutting down. I've had weird issues with the application no longer responding to events and I'm trying to shut it down and rebuild the event stream in this case but it just ends up that I have 2 streams listening for events and every other operation then can't get a connection from the pool and times out. It could be that the connection pool that this client creates need better configuration and then I wouldn't need to rebuild the event listener. Thoughts?

**Comment by bryanstephens** [comment]
I guess a work around could be to close the entire docker client and reinitialize one. I will probably roll with that.

**Comment by marcuslinke** [comment]
@bryanstephens Thanks for your feedback. 

1. I will make `ResultCallbackTemplate` an abstract class then
2. The default connection pool has only 2 connections configured so there probably will be problems when opening two event streams in parallel. Nevertheless the event stream connection should be released when calling `ResultCallback.close()`. Could you file a test case for the scenario you described please?

**Comment by marcuslinke** [comment]
Closed due to https://github.com/docker-java/docker-java/pull/263
